File: scala-client/src/main/scala/com/couchbase/client/scala/search/sort/SearchSort.scala
/*
 * Copyright (c) 2019 Couchbase, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.couchbase.client.scala.search.sort

import com.couchbase.client.core.api.search.queries.CoreGeoCoordinates
import com.couchbase.client.core.api.search.sort._

/** Base class for all FTS sort options in querying.
  *
  * @since 1.0.0
  */
sealed trait SearchSort {

  /** The identifier for the sort type, used in the "by" field.
    */
  protected def identifier: String

  protected def descending: Option[Boolean]

  private[scala] def toCore: CoreSearchSort
}

/** Base class for all FTS sort options in querying.
  *
  * @since 1.0.0
  */
object SearchSort {

  /** Sort by the hit score.
    *
    * @param descending whether the search results should be sorted in descending order.  If None (the default) is
    *                   specified, it is not sent and left to the server default.
    *
    * @since 1.0.0
    */
  case class ScoreSort(descending: Option[Boolean] = None) extends SearchSort {
    override protected def identifier = "score"

    override private[scala] def toCore =
      new CoreSearchSortScore(descending.map(_.asInstanceOf[java.lang.Boolean]).orNull)
  }

  /** Sort by the document ID.
    *
    * @param descending whether the search results should be sorted in descending order.  If None (the default) is
    *                   specified, it is not sent and left to the server default.
    *
    * @since 1.0.0
    */
  case class IdSort(descending: Option[Boolean] = None) extends SearchSort {
    override protected def identifier = "id"

    override private[scala] def toCore =
      new CoreSearchSortId(descending.map(_.asInstanceOf[java.lang.Boolean]).orNull)
  }

  /** Sort by a field in the rows.
    *
    * @param field      the name of the field to sort on
    * @param typ        the type of the field.  If left at the default `None`, it will default on the server to `Auto`
    * @param mode       the sort mode.  If left at the default `None`, it will default on the server to `Default`
    * @param missing    the missing mode.  If left at the default `None`, it will default on the server to `Last`
    * @param descending whether the search results should be sorted in descending order.  If None (the default) is
    *                   specified, it is not sent and left to the server default.
    */
  case class FieldSort(
      field: String,
      typ: Option[FieldSortType] = None,
      mode: Option[FieldSortMode] = None,
      missing: Option[FieldSortMissing] = None,
      descending: Option[Boolean] = None
  ) extends SearchSort {
    override protected def identifier = "field"

    override private[scala] def toCore =
      new CoreSearchSortField(
        field,
        typ.map {
          case FieldSortType.Auto   => CoreSearchFieldType.AUTO
          case FieldSortType.String => CoreSearchFieldType.STRING
          case FieldSortType.Number => CoreSearchFieldType.NUMBER
          case FieldSortType.Date   => CoreSearchFieldType.DATE
        }.orNull,
        mode.map {
          case FieldSortMode.Default => CoreSearchFieldMode.DEFAULT
          case FieldSortMode.Min     => CoreSearchFieldMode.MIN
          case FieldSortMode.Max     => CoreSearchFieldMode.MAX
        }.orNull,
        missing.map {
          case FieldSortMissing.Last  => CoreSearchFieldMissing.LAST
          case FieldSortMissing.First => CoreSearchFieldMissing.FIRST
        }.orNull,
        descending.map(_.asInstanceOf[java.lang.Boolean]).orNull
      )
  }

  /** Sort by a geo location distance.
    *
    * @param location   should be a two-value Seq containing the longitude and latitude (in that order) to centre the
    *                   search on
    * @param field      the name of the field to sort on
    * @param unit       the unit multipler to use for sorting.  Acceptable values are: "inch", "yards", "feet",
    *                   "kilometers", "nauticalmiles", "millimeters", "centimeters", "miles" and "meters".  Standard
    *                   abbreviations for those ("ft", "km", "mi") may also be used.  If left at default `None`, it
    *                   is not sent and left to the server default.
    * @param descending whether the search results should be sorted in descending order.  If None (the default) is
    *                   specified, it is not sent and left to the server default.
    *
    * @since 1.0.0
    */
  case class GeoDistanceSort(
      location: Seq[Float],
      field: String,
      descending: Option[Boolean] = None,
      unit: Option[String] = None
  ) extends SearchSort {

    override protected def identifier = "geo_distance"

    def unit(unit: String): GeoDistanceSort = {
      copy(unit = Some(unit))
    }

    override private[scala] def toCore =
      new CoreSearchSortGeoDistance(
        CoreGeoCoordinates.lon(location.head).lat(location(1)),
        field,
        unit
          .map(v =>
            v.toLowerCase match {
              case "inch" | "in" | "inches"             => CoreSearchGeoDistanceUnits.INCH
              case "yards"                              => CoreSearchGeoDistanceUnits.YARDS
              case "feet" | "ft"                        => CoreSearchGeoDistanceUnits.FEET
              case "kilometers" | "km" | "kilometres"   => CoreSearchGeoDistanceUnits.KILOMETERS
              case "nauticalmiles"                      => CoreSearchGeoDistanceUnits.NAUTICAL_MILES
              case "millimeters" | "mm" | "millimetres" => CoreSearchGeoDistanceUnits.MILLIMETERS
              case "centimeters" | "cm" | "centimetres" => CoreSearchGeoDistanceUnits.CENTIMETERS
              case "miles" | "mi"                       => CoreSearchGeoDistanceUnits.MILES
              case "meters" | "m" | "metres"            => CoreSearchGeoDistanceUnits.METERS
            }
          )
          .orNull,
        descending.map(_.asInstanceOf[java.lang.Boolean]).orNull
      )
  }

}

/** The type of the field used for a [[com.couchbase.client.scala.search.sort.SearchSort.FieldSort]]. */
sealed trait FieldSortType

object FieldSortType {
  case object Auto   extends FieldSortType
  case object String extends FieldSortType
  case object Number extends FieldSortType
  case object Date   extends FieldSortType
}

/** The mode of a [[com.couchbase.client.scala.search.sort.SearchSort.FieldSort]]. */
sealed trait FieldSortMode

object FieldSortMode {
  case object Default extends FieldSortMode
  case object Min     extends FieldSortMode
  case object Max     extends FieldSortMode
}

/** The missing behaviour for a [[com.couchbase.client.scala.search.sort.SearchSort.FieldSort]]. */
sealed trait FieldSortMissing

object FieldSortMissing {
  case object First extends FieldSortMissing
  case object Last  extends FieldSortMissing
}

File: kotlin-client/src/main/kotlin/com/couchbase/client/kotlin/search/SearchSort.kt
/*
 * Copyright 2022 Couchbase, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.couchbase.client.kotlin.search

import com.couchbase.client.core.api.search.sort.CoreSearchFieldMissing
import com.couchbase.client.core.api.search.sort.CoreSearchFieldMode
import com.couchbase.client.core.api.search.sort.CoreSearchFieldType
import com.couchbase.client.core.api.search.sort.CoreSearchSort
import com.couchbase.client.core.api.search.sort.CoreSearchSortField
import com.couchbase.client.core.api.search.sort.CoreSearchSortGeoDistance
import com.couchbase.client.core.api.search.sort.CoreSearchSortId
import com.couchbase.client.core.api.search.sort.CoreSearchSortScore
import com.couchbase.client.core.api.search.sort.CoreSearchSortString
import com.couchbase.client.kotlin.search.SearchSort.Companion.of

public enum class Direction { ASCENDING, DESCENDING }

public enum class FieldType(internal val core: CoreSearchFieldType) {
    AUTO(CoreSearchFieldType.AUTO),
    STRING(CoreSearchFieldType.STRING),
    NUMBER(CoreSearchFieldType.NUMBER),
    DATE(CoreSearchFieldType.DATE),
    ;
}

public enum class Mode(internal val core: CoreSearchFieldMode) {
    DEFAULT(CoreSearchFieldMode.DEFAULT),
    MIN(CoreSearchFieldMode.MIN),
    MAX(CoreSearchFieldMode.MAX),
    ;
}

public enum class Missing(internal val core: CoreSearchFieldMissing) {
    FIRST(CoreSearchFieldMissing.FIRST),
    LAST(CoreSearchFieldMissing.LAST),
    ;
}

/**
 * Specifies how to sort the results of a search query.
 *
 * Create instances using the companion factory methods.
 *
 * For tiered sorting, chain using [then] or [of].
 *
 * @sample com.couchbase.client.kotlin.samples.searchTieredSort
 */
public class SearchSort internal constructor(private val components: List<SearchSortComponent>) {
    internal constructor(e: SearchSortComponent) : this(listOf(e))

    internal val core : List<CoreSearchSort> = components.map { it.core }


    /**
     * Returns a tiered sort consisting of this sort followed by [other].
     * @sample com.couchbase.client.kotlin.samples.searchTieredSort
     */
    public infix fun then(other: SearchSort) : SearchSort = SearchSort(components + other.components)

    override fun toString(): String {
        return "SearchSort($components)"
    }

    public companion object {

        public fun byScore(
            direction: Direction = Direction.DESCENDING,
        ): SearchSort = SearchSort(SearchSortScore(direction))

        public fun byId(
            direction: Direction = Direction.ASCENDING,
        ): SearchSort = SearchSort(SearchSortId(direction))

        /**
         * @param field Specifies the name of a field on which to sort.
         *
         * @param mode Specifies the search-order for index-fields that
         * contain multiple values (as a consequence of arrays or multi-token
         * analyzer-output). The default order is undefined but deterministic,
         * allowing the paging of results from `from (offset)` with reliable
         * ordering. To sort using the minimum or maximum value, the value
         * of mode should be set to either min or max.
         *
         * @param type Specifies the type of the search-order field value.
         * For example, string for text fields, date for DateTime fields,
         * or number for numeric/geo fields.
         *
         * @param missing Specifies the sort-procedure for documents with
         * a missing value in a field specified for sorting. The value of
         * missing can be first, in which case results with missing values
         * appear before other results, or last (the default), in which case
         * they appear after.
         */
        public fun byField(
            field: String,
            type: FieldType = FieldType.AUTO,
            mode: Mode = Mode.DEFAULT,
            missing: Missing = Missing.LAST,
            direction: Direction = Direction.ASCENDING,
        ): SearchSort = SearchSort(SearchSortField(field, type, mode, missing, direction))

        public fun byGeoDistance(
            field: String,
            location: GeoPoint,
            unit: GeoDistanceUnit = GeoDistanceUnit.METERS,
            direction: Direction = Direction.ASCENDING,
        ): SearchSort = SearchSort(SearchSortGeoDistance(field, location, unit, direction))

        /**
         * Specifies the sort order using strings.
         * Syntax is described by [Sorting With Strings](https://docs.couchbase.com/server/current/fts/fts-search-response.html#sorting-with-strings)
         */
        public fun by(firstSort: String, vararg remainingSorts: String): SearchSort =
            by(listOf(firstSort, *remainingSorts))

        /**
         * Specifies the sort order using strings.
         * Syntax is described by [Sorting With Strings](https://docs.couchbase.com/server/current/fts/fts-search-response.html#sorting-with-strings)
         */
        public fun by(sorts: List<String>): SearchSort =
            SearchSort(sorts.map { SearchSortFromString(it) })

        /**
         * Concatenates [sorts] into a single SearchSort,
         * for when it's convenient to build a sort from a list
         * instead of using [then].
         */
        public fun of(sorts: List<SearchSort>): SearchSort = SearchSort(sorts.flatMap { it.components })
    }
}

internal sealed class SearchSortComponent {
    internal abstract val core: CoreSearchSort
    override fun toString(): String = core.toString()
}

internal class SearchSortFromString internal constructor(sort: String) : SearchSortComponent() {
    override val core = CoreSearchSortString(sort)
}

internal class SearchSortScore internal constructor(direction: Direction) : SearchSortComponent() {
    override val core = CoreSearchSortScore(direction == Direction.DESCENDING)
}

internal class SearchSortId internal constructor(direction: Direction) : SearchSortComponent() {
    override val core = CoreSearchSortId(direction == Direction.DESCENDING)
}

internal class SearchSortField internal constructor(
    field: String,
    type: FieldType,
    mode: Mode,
    missing: Missing,
    direction: Direction,
) : SearchSortComponent() {
    override val core = CoreSearchSortField(
        field,
        type.core,
        mode.core,
        missing.core,
        direction == Direction.DESCENDING
    )
}

internal class SearchSortGeoDistance internal constructor(
    private val field: String,
    private val location: GeoPoint,
    private val unit: GeoDistanceUnit,
    private val direction: Direction,
) : SearchSortComponent() {
    override val core = CoreSearchSortGeoDistance(
        location.core,
        field,
        unit.core,
        direction == Direction.DESCENDING,
    )
}
